File: src/server/aggregators/base/utils/youtube.ts
/**
 * YouTube utility functions.
 */

import { logger } from "@server/utils/logger";

/**
 * Extract YouTube video ID from URL.
 */
export function extractYouTubeVideoId(url: string): string | null {
  try {
    const parsedUrl = new URL(url);
    let videoId: string | null = null;

    // Handle youtu.be short URLs
    if (
      parsedUrl.hostname === "youtu.be" ||
      parsedUrl.hostname === "www.youtu.be"
    ) {
      videoId = parsedUrl.pathname.slice(1).split("?")[0].split("&")[0];
    }
    // Handle youtube.com URLs
    else if (
      parsedUrl.hostname === "youtube.com" ||
      parsedUrl.hostname === "www.youtube.com" ||
      parsedUrl.hostname === "m.youtube.com"
    ) {
      // /watch?v=VIDEO_ID
      if (parsedUrl.pathname === "/watch" || parsedUrl.pathname === "/watch/") {
        videoId = parsedUrl.searchParams.get("v");
      }
      // /embed/VIDEO_ID or /v/VIDEO_ID or /shorts/VIDEO_ID
      else if (
        parsedUrl.pathname.startsWith("/embed/") ||
        parsedUrl.pathname.startsWith("/v/") ||
        parsedUrl.pathname.startsWith("/shorts/")
      ) {
        const parts = parsedUrl.pathname.split("/");
        if (parts.length > 2) {
          videoId = parts[2].split("?")[0];
        }
      }
    }

    // Validate video ID format (typically 11 characters, alphanumeric with - and _)
    if (videoId && /^[\w-]+$/.test(videoId)) {
      return videoId;
    }

    return null;
  } catch (error) {
    logger.debug({ error, url }, "Failed to extract YouTube video ID");
    return null;
  }
}

/**
 * Get YouTube proxy URL for embedding.
 * Always uses BASE_URL from environment if set, otherwise defaults to frontend port (4200) in development
 * or backend port (3000) in production.
 */
export function getYouTubeProxyUrl(videoId: string): string {
  const baseUrl =
    process.env["BASE_URL"] ||
    (process.env["NODE_ENV"] === "development"
      ? "http://localhost:4200"
      : "http://localhost:3000");
  return `${baseUrl.replace(/\/$/, "")}/api/youtube-proxy?v=${encodeURIComponent(videoId)}`;
}

/**
 * Create YouTube embed HTML using proxy URL and embed container class.
 * Returns HTML string that can be used directly or loaded into cheerio.
 *
 * @param videoId - YouTube video ID
 * @param caption - Optional caption HTML to append after the iframe
 * @returns HTML string with youtube-embed-container div and iframe
 */
export function createYouTubeEmbedHtml(
  videoId: string,
  caption?: string,
): string {
  const embedUrl = getYouTubeProxyUrl(videoId);
  const iframeHtml =
    `<div class="youtube-embed-container">` +
    `<iframe src="${embedUrl}" ` +
    `title="YouTube video player" ` +
    `frameborder="0" ` +
    `style="width: 100%" ` +
    `allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" ` +
    `allowfullscreen></iframe>` +
    (caption ? caption : "") +
    `</div>`;
  return iframeHtml;
}

File: core/aggregators/utils/youtube.py
"""
YouTube utilities for header element extraction.

Provides functions for:
- Detecting and extracting YouTube video IDs
- Generating YouTube embed HTML
- Constructing thumbnail URLs
"""

import re
from typing import Optional

from bs4 import BeautifulSoup
from django.conf import settings


def extract_youtube_video_id(url: str) -> Optional[str]:
    """
    Extract YouTube video ID from various URL formats.

    Handles:
    - youtu.be/{ID}
    - youtube.com/watch?v={ID}
    - youtube.com/embed/{ID}
    - youtube.com/v/{ID}
    - youtube.com/shorts/{ID}

    Args:
        url: YouTube URL in various formats

    Returns:
        Video ID if valid format found, None otherwise
    """
    if not url:
        return None

    patterns = [
        # youtu.be short URL
        r"youtu\.be/([A-Za-z0-9_-]+)",
        # youtube.com watch URL
        r"youtube\.com/watch\?.*v=([A-Za-z0-9_-]+)",
        # youtube.com embed URL
        r"youtube\.com/embed/([A-Za-z0-9_-]+)",
        # youtube.com /v/ URL
        r"youtube\.com/v/([A-Za-z0-9_-]+)",
        # youtube.com shorts
        r"youtube\.com/shorts/([A-Za-z0-9_-]+)",
    ]

    for pattern in patterns:
        match = re.search(pattern, url)
        if match:
            video_id = match.group(1)
            # Validate video ID format
            if re.match(r"^[A-Za-z0-9_-]{11}$", video_id) or re.match(
                r"^[A-Za-z0-9_-]+$", video_id
            ):
                return video_id

    return None


def get_youtube_thumbnail_url(video_id: str, quality: str = "maxresdefault") -> str:
    """
    Get YouTube thumbnail URL for a video.

    Quality options (in order of preference):
    - maxresdefault: Highest quality (1280x720)
    - hqdefault: High quality (480x360)
    - sddefault: Standard quality (640x480)
    - mqdefault: Medium quality (320x180)
    - default: Default (120x90)

    Args:
        video_id: YouTube video ID
        quality: Thumbnail quality level

    Returns:
        URL to thumbnail image
    """
    return f"https://img.youtube.com/vi/{video_id}/{quality}.jpg"


def get_youtube_proxy_url(video_id: str) -> str:
    """
    Get full YouTube proxy URL for a video.

    Args:
        video_id: YouTube video ID

    Returns:
        Full URL to the proxy endpoint
    """
    return f"{settings.BASE_URL}/api/youtube-proxy?v={video_id}"


def create_youtube_embed_html(video_id: str, caption: str = "") -> str:
    """
    Create HTML for embedded YouTube video.

    Generates an iframe element that uses a full proxy endpoint for embedding
    (to avoid embedding YouTube's standard iframe which may have
    privacy/tracking considerations).

    Args:
        video_id: YouTube video ID
        caption: Optional caption to append after iframe

    Returns:
        HTML string with youtube-embed-container div and iframe
    """
    proxy_url = get_youtube_proxy_url(video_id)

    html = (
        f'<div class="youtube-embed-container">'
        f'<iframe src="{proxy_url}" '
        f'title="YouTube video player" '
        f'width="560" '
        f'height="315" '
        f'frameborder="0" '
        f'scrolling="no" '
        f"allowfullscreen></iframe>"
    )

    if caption:
        html += caption

    html += "</div>"

    return html


def is_youtube_url(url: str) -> bool:
    """
    Check if a URL is a YouTube URL.

    Args:
        url: URL to check

    Returns:
        True if URL is from youtube.com or youtu.be
    """
    if not url:
        return False

    youtube_domains = ["youtube.com", "youtu.be", "m.youtube.com", "youtube-nocookie.com"]
    return any(domain in url for domain in youtube_domains)


def proxy_youtube_embeds(soup: BeautifulSoup) -> None:
    """
    Find and replace YouTube iframes with proxy embeds.

    Args:
        soup: BeautifulSoup object to modify in-place
    """
    for iframe in soup.find_all("iframe"):
        src = iframe.get("src", "")
        if not src:
            continue

        if is_youtube_url(src):
            video_id = extract_youtube_video_id(src)
            if video_id:
                # Create replacement HTML
                replacement_html = create_youtube_embed_html(video_id)
                replacement_soup = BeautifulSoup(replacement_html, "html.parser")

                # The first element should be the div
                new_tag = replacement_soup.find("div", class_="youtube-embed-container")
                if new_tag:
                    iframe.replace_with(new_tag)
